core/src/main/java/com/kin/ecosystem/core/data/blockchain/BlockchainSourceImpl.java

```diff
@@ -101,24 +108,81 @@ public static BlockchainSourceImpl getInstance() {
 	}
 
 	@Override
-	public void createAccount() throws BlockchainException {
-		createKinAccountIfNeeded();
+	public void loadAccount(String kinUserId) throws BlockchainException {
+		migrateToMultipleUsers(kinUserId);
+		createOrLoadAccount(kinUserId);
 		initBalance();
 	}
 
-	private void createKinAccountIfNeeded() throws BlockchainException {
-		if (kinClient.hasAccount()) {
+	/**
+	 * Support backward compatibility,
+	 * load the old account and migrate to current multiple users implementation.
+	 */
+	private void migrateToMultipleUsers(String kinUserId) {
+		final int accountIndex = local.getAccountIndex();
+		if (accountIndex != NOT_EXIST && kinClient.hasAccount()) {
+			Logger.log(new Log().withTag(TAG).put("migrateToMultipleUsers user:", kinUserId));
+			KinAccount account = kinClient.getAccount(accountIndex);
+			local.setActiveUserWallet(kinUserId, account.getPublicAddress());
+			local.deleteAccountIndexKey();
+		}
+	}
+
+	private void createOrLoadAccount(String kinUserId) throws BlockchainException {
+		Set<String> wallets = local.getUserWallets(kinUserId);
+		if (kinClient.hasAccount() && wallets != null && wallets.size() > 0) {
+			Logger.log(new Log().withTag(TAG).text("createOrLoadAccount").put("currentUserId", currentUserId)
+				.put("kinUserId", kinUserId));
+			final String lastWalletAddress;
+			if (!StringUtil.isEmpty(currentUserId)) {
+				if (currentUserId.equals(kinUserId)) {
+					// Get last active wallet from cache
+					lastWalletAddress = local.getCurrentWalletAddress();
+				} else {
+					// Get last wallet added
+					lastWalletAddress = getLastWalletAddress(wallets);
+					this.currentUserId = kinUserId;
+				}
+
+				// Match between last wallet address to wallets on device.
+				if (!StringUtil.isEmpty(lastWalletAddress)) {
+					for (int i = 0; i < kinClient.getAccountCount(); i++) {
+						KinAccount account = kinClient.getAccount(i);
+						if (lastWalletAddress.equals(account.getPublicAddress())) {
+							this.account = account;
+							break;
+						}
+					}
+				}
+			}
+
+			// No matching found
 			if (account == null) {
-				account = kinClient.getAccount(activeAccountIndex);
+				account = createAccount();
 			}
+
 		} else {
-			try {
-				// Create new account
-				account = kinClient.addAccount();
-				setAccountIndex(0);
-			} catch (CreateAccountException e) {
-				throw ErrorUtil.getBlockchainException(e);
-			}
+			account = createAccount();
+		}
+		local.setActiveUserWallet(kinUserId, account.getPublicAddress());
+	}
+
+	@Nullable
+	private String getLastWalletAddress(Set<String> wallets) {
```

Comment:
getLastWalletAddress should be part of local and probably we can make getUserWallets private.

Reply:
I don't think it should be part of local, because it's the repo logic.
the local can supply the **current** last active wallet and getUserWallets nothing more.
to get the last from the user wallets is the repo responsibility.